### ISSUE TITLE
Alternative to `np.copy`

### ISSUE THREAD

**Issue by harryjulian** [ml-explore/mlx#279]
Running into an issue where I need to copy an array -- no mention of a `.copy` or `.clone` method when searching the docs. Seems to exist but not in the python bindings. Any alternatives at the moment, or is this another matter of needs to be implemented under the hood?

Example:

```python

def transform(X, transform):
    ...
    return X_transformed

X = load_data()
X_a = transform(X, a)
X_b = transform(X, b)
# X_a == X_b
```

**Comment by awni** [comment]
Could you elaborate on your use case?  Where do you need the copy?  In you're example if `a` and `b` are operating on `X` then you should not have `X_a` and `X_b` as the same array

**Comment by Viditagarwal7479** [comment]
Here you don't need `.clone` method unlike framework like PyTorch as here by default a copy of is created.

For example:
```py
import mlx.core as mx
a = mx.array([[1,2],[3,4]])
b = a
b *= 2
print(b)
# array([[2, 4], [6, 8]], dtype=int32)
print(a)
# array([[1, 2], [3, 4]], dtype=int32)
```

**Comment by awni** [comment]
Closing this, let us know if you still have questions about it!

**Comment by JoeJoe1313** [comment]
Hello, expanding the initial question: If I have a case where I need to do some iterative updates as in this case:

```
v_old = np.copy(v)
c_old = np.copy(c)
```

Am I correct that the way to do this in `mlx` would be

```
v_old = mx.array(v)
c_old = mx.array(c)
```

**Comment by awni** [comment]
Could you provide a bit more context? I don't think you need that `mx.array(v)` in general. But if you are doing something in-place then maybe you do.

For example:

```python
v = mx.array([1])
prev = v
v = v + 1
# prev is still 1
```

```python
v = mx.array([1])
prev = v
v += 1
# prev is now 2
```

**Comment by JoeJoe1313** [comment]
Yes, exactly, that was the example I was aiming for. So, if I am doing something in-place using `mx.array(x)` should be the solution at the moment? Are you planning on implementing `.copy` or `.clone` at some point in the future?

**Comment by awni** [comment]
We don't really have a plan to implement either. Do you prefer `.copy` over `mx.array`? Could you say more about why?

**Comment by laclouis5** [comment]
I have a use case where I want to copy the weights from one module to a new module while keeping both of them around, independently.

What is recommended way to achieve that in the absence of `.copy` method?

**Comment by awni** [comment]
```python
old_module = nn.Linear(3, 3)
new_module = nn.Linear(3, 3)
new_module.update(old_module.parameters())
old_module.weight = mx.array([1,2,3])
print(old_module.weight)
print(new_module.weight)
```

**Comment by laclouis5** [comment]
@awni Thanks but it does not seem to do a copy. 

The piece of code you show assigns new weights to the new module while I want the *same* weights as the old module. Also, the bias is still share between the old and new module.

I took a look at the `.update` code and it does not seem to to a copy the data unfortunately.

**Comment by awni** [comment]
Can you be more specific about the behavior you want? In the snippet above after you do the update the two modules share the same weights.

**Comment by awni** [comment]
Your code snipped is the same as the `update`, a copy is made. But if you write to the `new_module.weight` then it won't be reflected in the `old_module.weight`.

You can copy all the parameters from one module to another like this:

```
new_module.update(old_module.parameters())
```

If you just want a subset of the parameters then filter `old_module.parameters()` before calling `new_module.update` or use [`filter_and_map`](https://ml-explore.github.io/mlx/build/html/python/nn/_autosummary/mlx.nn.Module.filter_and_map.html#mlx.nn.Module.filter_and_map).

**Comment by laclouis5** [comment]
They don't seem to share the same weights:

```python
old_module = nn.Linear(3, 3)
new_module = nn.Linear(3, 3)
new_module.weight = mx.array(old_module.weight)

old_module.weight[:, :] = 0
print(old_module.weight, new_module.weight)
```

Here, a copy seems to be made since only the old module has zero weights. With your suggestion based on `.update`, both modules have zero weights.

However, this does not appear to pose any issue during training. In my tests, `opt.update()` seems to make a copy somewhere and only the old model weight is updated but I'm not entirely sure.

The behavior I'm looking for is simple: I want a copy of an array, i.e a new array with the same contents as the original one but without referencing it in any way (computational graph, gradients, computation, etc.).

I need this feature to duplicate, modify and convert modules between multiple forms.

**Comment by awni** [comment]
Does this work for you?

```
from copy import deepcopy

old_module = nn.Linear(3, 3)
new_module = nn.Linear(3, 3)
new_module.update(deepcopy(old_module.parameters()))
```

**Comment by laclouis5** [comment]
This also seems to work yes.
